src/coupled_driver.cpp

```diff
@@ -191,132 +226,213 @@ void CoupledDriver::compute_temperature_norm(const CoupledDriver::Norm& n,
     std::copy(temperatures_.begin(), temperatures_.end(), temperatures_prev_.begin());
   }
 
-  // Compute the next iterate of the temperature
-  auto& heat = get_heat_driver();
-  if (heat.active()) {
-    auto t = heat.temperature();
+  temperatures_ = heat.temperature();
+  this->comm_.send_and_recv(temperatures_, coupling_root_, heat_root_);
 
-    if (heat.has_coupling_data())
-      temperatures_ = alpha_T_ * t + (1.0 - alpha_T_) * temperatures_prev_;
+  if (relax && comm_.rank == coupling_root_) {
+    temperatures_ = alpha_T_ * temperatures_ + (1.0 - alpha_T_) * temperatures_prev_;
   }
 
-  // Set temperature in the neutronics solver
-  set_temperature();
+  // ****************************************************************************
+  // Send underrelaxed temperature to all neutron ranks and set cell temperatures
+  // ****************************************************************************
+
+  // Broadcast global_element_temperatures onto all the neutronics procs
+  comm_.send_and_recv(temperatures_, neutronics_root_, coupling_root_);
+  neutronics.comm_.broadcast(temperatures_);
+
+  if (neutronics.active()) {
+    // For each neutronics cell, volume average temperatures and set
+    for (CellHandle cell = 0; cell < cell_to_elems_.size(); ++cell) {
+      // Get volume-average temperature for this cell instance
+      double average_temp = 0.0;
+      double total_vol = 0.0;
+      for (int elem : cell_to_elems_.at(cell)) {
+        double T = temperatures_[elem];
+        double V = elem_volumes_[elem];
+        average_temp += T * V;
+        total_vol += V;
+      }
+      // Set temperature for cell instance
+      average_temp /= total_vol;
+      Ensures(average_temp > 0.0);
+      neutronics.set_temperature(cell, average_temp);
+    }
+  }
 }
 
-void CoupledDriver::update_density()
+void CoupledDriver::update_density(bool relax)
 {
-  // Store previous density solution; a previous solution will always be present
-  // because a density IC is set and the neutronics solver runs first
-  if (has_global_coupling_data()) {
+
+  auto& neutronics = this->get_neutronics_driver();
+  auto& heat = this->get_heat_driver();
+
+  // *********************************************************************
+  // Send heat driver's density to coupling root and apply underrelaxation
+  // *********************************************************************
+
+  if (relax && comm_.rank == coupling_root_) {
     std::copy(densities_.begin(), densities_.end(), densities_prev_.begin());
   }
 
-  auto& heat = get_heat_driver();
-  if (heat.active()) {
-    auto d = heat.density();
+  densities_ = heat.density();
+  comm_.send_and_recv(densities_, coupling_root_, heat_root_);
 
-    if (heat.has_coupling_data())
-      densities_ = alpha_rho_ * d + (1.0 - alpha_rho_) * densities_prev_;
+  if (relax && comm_.rank == coupling_root_) {
+    densities_ = alpha_rho_ * densities_ (1.0 - alpha_rho_) * densities_prev_;
```

Comment:
This is causing the surrogate to fail currently
```suggestion
    densities_ = alpha_rho_ * densities_ + (1.0 - alpha_rho_) * densities_prev_;
```

Reply:
Thanks!